### ISSUE TITLE
Trying to train a Custom FHog Like Descriptor to run in external library

### ISSUE THREAD

**Issue by StanJBrown** [davisking/dlib#2018]
## Expected Behavior
I am trying to train a hog-like descriptor using the object detector / fhog_scan_pyramid. 
( I followed these instructions here:[https://github.com/davisking/dlib/blob/6b581d91f6b9b847a8163420630ef947e7cc88db/dlib/image_processing/scan_fhog_pyramid.h#L18]()

and here:[http://dlib.net/fhog_object_detector_ex.cpp.html]( )
The SVM model is then run on piece of hardware that has accelerated support for a hog like feature  (it cannot support FHOG, otherwise I would just use that)

## Current Behavior
In dlib and with testing the descriptor works fine and shows an amazing detection rate and low false positives. When I then take the w vector from the resultant object detector, using this call: [https://github.com/davisking/dlib/blob/6b581d91f6b9b847a8163420630ef947e7cc88db/dlib/image_processing/object_detector.h#L104]() and move it over to my own software  I get vastly different results. At first I failed to account for the padding and was getting out a w that was longer then expected (I missed the part about the 1 cell padding around the search region window size), but now I get out a feature vector of the expected length. 

My question is, from my understanding, when the training / object detection is run, a window is slid across the HOG image (basically an image containing cells of hog-like feature vectors) and the descriptor returned is a row-major vector of each descriptor in the window concatenated with each cell concatenated into a single vector (row wise). I should in theory then be able to in my own code multiply my descriptors with the svm model (w() param in the dlib object detector). 

I am currently trying to understand if I am missing something in the dectect() function call in the object detector and the hog filter-banks that are called in that piece of code. Are there some crucial parts of code I am missing when trying to port this SVM model to my own library? I assumed that this code is an optimization step that speeds up detection, but please correct me if I am wrong. 

## Steps to Reproduce
Train a custom, hog-like descriptor and then try to use it somewhere else assuming the svm vector from get_w() is row major and the svm score is equal to your descriptor * svm_vector

* **Version**:  Pulled the current master
* **Where did you get dlib**: github
* **Platform**: Ubuntu 18.04
* **Compiler**: gcc

**Comment by davisking** [comment]
The HOG features are represented as a multi-channel planar image.  Detection just convolves (really it does a cross correlation, but for some reason image processing literature usually uses the word convolve) that image with a learned filter.  It's not doing any funny stuff.

How that hog vector is laid out as a multi-channel image is discussed in the documentation.  Did you see https://github.com/davisking/dlib/blob/6b581d91f6b9b847a8163420630ef947e7cc88db/dlib/image_processing/scan_fhog_pyramid_abstract.h?

For example, a 31 channel HOG image gets turned into a feature vector by treating each channel as a vector (by reading it in row major order) and then concatenating them.  It sounds like you are not setting it up that way.  Think of the cells as channels in a multi-channel image.  To say this another way, all the features from channel `i` are adjacent in the `w` vector.

**Comment by StanJBrown** [comment]
Thanks for the quick reply. 
Thanks for the clarification. I had misinterpreted the layout of the vector completely, so I wrote a script to remap it to what I am expecting and my results are now a fair bit closer to what I would expect. I might be having an over-fitting problem now as my descriptor has 500+ elements compared to the 31 of FHOG so I will work at fixing that (I am only training with about 200 images at the moment with 250 positive boxes).

And yes, I was using [https://github.com/davisking/dlib/blob/6b581d91f6b9b847a8163420630ef947e7cc88db/dlib/image_processing/scan_fhog_pyramid_abstract.h](https://github.com/davisking/dlib/blob/6b581d91f6b9b847a8163420630ef947e7cc88db/dlib/image_processing/scan_fhog_pyramid_abstract.h). It was pretty clear about how to set it up, except for my brain-fart / misinterpretation of how the vector was laid out.
